benchmark/spmv/cuda_linops.hpp

```diff
@@ -0,0 +1,484 @@
+
+public:
+    using csr = gko::matrix::Csr<ValueType, IndexType>;
+    using mat_data = gko::matrix_data<ValueType, IndexType>;
+
+    void read(const mat_data &data) override
+    {
+        csr_->read(data);
+        this->set_size(gko::dim<2>{csr_->get_size()});
+    }
+
+    ~CuspCsrEx() override
+    {
+        const auto id = this->get_gpu_exec()->get_device_id();
+        gko::device_guard g{id};
+        if (set_buffer_) {
+            try {
+                GKO_ASSERT_NO_CUDA_ERRORS(cudaFree(buffer_));
+            } catch (const std::exception &e) {
+                std::cerr
+                    << "Error when unallocating CuspCsrEx temporary buffer: "
+                    << e.what() << std::endl;
+            }
+        }
+    }
+
+protected:
+    void apply_impl(const gko::LinOp *b, gko::LinOp *x) const override
+    {
+        auto dense_b = gko::as<gko::matrix::Dense<ValueType>>(b);
+        auto dense_x = gko::as<gko::matrix::Dense<ValueType>>(x);
+        auto db = dense_b->get_const_values();
+        auto dx = dense_x->get_values();
+        ValueType alpha = gko::one<ValueType>();
+        ValueType beta = gko::zero<ValueType>();
+        gko::size_type buffer_size = 0;
+        auto data_type = gko::kernels::cuda::cuda_data_type<ValueType>();
+
+        const auto id = this->get_gpu_exec()->get_device_id();
+        gko::device_guard g{id};
+        // This function seems to require the pointer mode to be set to HOST.
+        // Ginkgo use pointer mode DEVICE by default, so we change this
+        // temporarily.
+        auto handle = this->get_gpu_exec()->get_cusparse_handle();
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(
+            cusparseSetPointerMode(handle, CUSPARSE_POINTER_MODE_HOST));
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(cusparseCsrmvEx_bufferSize(
+            handle, algmode_, trans_, this->get_size()[0], this->get_size()[1],
+            csr_->get_num_stored_elements(), &alpha, data_type,
+            this->get_descr(), csr_->get_const_values(), data_type,
+            csr_->get_const_row_ptrs(), csr_->get_const_col_idxs(), db,
+            data_type, &beta, data_type, dx, data_type, data_type,
+            &buffer_size));
+        GKO_ASSERT_NO_CUDA_ERRORS(cudaMalloc(&buffer_, buffer_size));
+        set_buffer_ = true;
+
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(cusparseCsrmvEx(
+            handle, algmode_, trans_, this->get_size()[0], this->get_size()[1],
+            csr_->get_num_stored_elements(), &alpha, data_type,
+            this->get_descr(), csr_->get_const_values(), data_type,
+            csr_->get_const_row_ptrs(), csr_->get_const_col_idxs(), db,
+            data_type, &beta, data_type, dx, data_type, data_type, buffer_));
+
+        // Set the pointer mode back to the default DEVICE for Ginkgo
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(
+            cusparseSetPointerMode(handle, CUSPARSE_POINTER_MODE_DEVICE));
+    }
+
+
+    gko::size_type get_num_stored_elements() const noexcept
```

Comment:
Not sure if you also wanted to make this function `protected` since it is `public` in class `CuspCsrmm`.